skyportal/handlers/api/candidate.py

```diff
@@ -0,0 +1,280 @@
+        ):
+            return self.error("Page number out of range.")
+        info["candidates"] = (
+            q.limit(CANDIDATES_PER_PAGE).offset((page - 1) * CANDIDATES_PER_PAGE).all()
+        )
+
+        info["pageNumber"] = page
+        info["lastPage"] = info["totalMatches"] <= page * CANDIDATES_PER_PAGE
+        info["candidateNumberingStart"] = (page - 1) * CANDIDATES_PER_PAGE + 1
+        info["candidateNumberingEnd"] = min(
+            info["totalMatches"], page * CANDIDATES_PER_PAGE
+        )
+        if info["totalMatches"] == 0:
+            info["candidateNumberingStart"] = 0
+
+        return self.success(data=info)
+
+    @permissions(["Upload data"])
+    def post(self):
+        """
+        ---
+        description: Upload a candidate. If group_ids is not specified, the user or token's groups will be used.
+        requestBody:
+          content:
+            application/json:
+              schema: Source
+        responses:
+          200:
+            content:
+              application/json:
+                schema:
+                  allOf:
+                    - Success
+                    - type: object
+                      properties:
+                        id:
+                          type: string
+                          description: New candidate ID
+        """
+        data = self.get_json()
+        data["is_candidate"] = True
+        saved_as_source_by_id = data.pop("saved_as_source_by_id", None)
+        schema = Source.__schema__()
+        user_group_ids = [g.id for g in self.current_user.groups]
+        if not user_group_ids:
+            return self.error(
+                "You must belong to one or more groups before you can add candidates."
+            )
+        try:
+            group_ids = [id for id in data.pop("group_ids") if id in user_group_ids]
+        except KeyError:
+            group_ids = user_group_ids
+        if not group_ids:
+            return self.error(
+                "Invalid group_ids field. Please specify at least "
+                "one valid group ID that you belong to."
+            )
+        try:
+            c = schema.load(data)
+        except ValidationError as e:
+            return self.error(
+                "Invalid/missing parameters: " f"{e.normalized_messages()}"
+            )
+        groups = Group.query.filter(Group.id.in_(group_ids)).all()
+        if not groups:
+            return self.error(
+                "Invalid group_ids field. Please specify at least "
+                "one valid group ID that you belong to."
+            )
+        c.candidate_groups = groups
+        if saved_as_source_by_id is None:
+            saved_as_source_by_id = self.current_user.id
+        # TODO - create GroupSources with appropriate fields
+        DBSession.add(c)
+        DBSession().commit()
+
+        self.push_all(action="skyportal/FETCH_CANDIDATES")
+        return self.success(data={"id": c.id})
+
+    @permissions(["Manage sources"])
```

Comment:
The permissioning around deletion of a source vs deletion of a candidate will be very different. We basically will want almost no one to be able to delete a source (once it's official, it's got an official name and there's no reason to kill it even if it turns out to be bogus). Default should be only `superadmin` here. But many more roles should be able to delete candidates.

Reply:
I don't see this endpoint actually getting hit, as I *think* the only way candidates are completely wiped from the DB is when the vacuum cron job sweeps up old, unsaved candidates. Is this right? Or can a user delete a candidate so that it's no longer in the system at all?